common/date_utils.py

```diff
@@ -1,93 +1,121 @@
     "utc_time_max",
+    "utc_start_of_day",
     "combine",
-    "datetimes_equal",
-    "times_equal",
     "timedelta_to_json",
     "timedelta_from_json",
+    "time_as_timedelta",
 ]
 
 
-### LOCAL TIME ###########################################################################################
+DEFAULT_TIMEZONE = get_default_timezone()
 
 
-def local_timezone(*, ref: datetime.datetime | None = None) -> datetime.timezone:
-    """
-    Fetch local timezone used to convert naive datetimes to aware datetimes.
+class _TZComparator(Generic[TCanHaveTZ, TValid]):  # noqa: N801, RUF100
+    def __init__(self, _input_1: TCanHaveTZ, _input_2: TCanHaveTZ):
+        self._input_1 = self._validate_input(_input_1, name="Input 1")
+        self._input_2 = self._validate_input(_input_2, name="Input 2")
 
-    Cannot use `django.utils.timezone.get_default_timezone()`,
-    because it returns a `zoneinfo.ZoneInfo` object, which when used together with
-    `datetime.timezone` (e.g. datetime.UTC/datetime.timezone.utc) will not work as
-    expected when comparing with `datetime.time` objects:
+    def __bool__(self):
+        msg = f"Cannot compare '{self.__class__.__name__}' object directly. Did you forget to call a comparison method?"
+        raise RuntimeError(msg)
 
-    ```python 3.11
-    import zoneinfo
-    import datetime
+    @staticmethod
+    def _validate_input(_input: TCanHaveTZ, *, name: str) -> TValid:
+        raise NotImplementedError
 
-    tz = zoneinfo.ZoneInfo("Europe/Helsinki")
+    def is_lte(self) -> bool:
+        return self._input_1 <= self._input_2
 
-    dt_1 = datetime.datetime(2021, 1, 1, tzinfo=tz)
-    dt_2 = datetime.datetime(2021, 1, 1, tzinfo=datetime.UTC)
+    def is_gte(self) -> bool:
+        return self._input_1 >= self._input_2
 
-    assert dt_1 < dt_2  # Works as expected
+    def is_lt(self) -> bool:
+        return self._input_1 < self._input_2
 
-    t_1 = datetime.time(tzinfo=tz)
-    t_2 = datetime.time(tzinfo=datetime.UTC)
+    def is_gt(self) -> bool:
+        return self._input_1 > self._input_2
 
-    assert t_1 < t_2  # raises "TypeError: can't compare offset-naive and offset-aware times"
-    ```
-    """
-    return timezone_from_name(settings.TIME_ZONE, ref=ref)
+    def is_equal(self) -> bool:
+        return self._input_1 == self._input_2
 
+    def is_not_equal(self) -> bool:
+        return self._input_1 != self._input_2
 
-@singledispatch
-def as_local_timezone(
-    _input: datetime.datetime | datetime.time,
-    /,
-    *,
-    ref: datetime.datetime | None = None,
-) -> datetime.datetime | datetime.time:
-    """
-    Convert a datetime to local timezone.
 
-    Use this instead of `datetime.astimezone()` or `datetime.datetime(..., tzinfo=local_timezone())`
-    to avoid issues with daylight savings time.
-    """
-    raise TypeError(f"Unsupported type: {_input.__class__.__name__}")
+class compare_datetimes(_TZComparator[datetime.datetime, datetime.datetime]):  # noqa: N801, RUF100
```

Comment:
This is not actually used anywhere (yet?).

Reply:
Not yet, should probably use it inside the TimeSpanElement class.

Reply:
Actually, I don't think there is a need for it currently, as lt/lte/gt/gte comparisons already work correctly for datetime-objects, and there aren't any (not) equals checks. Code is simpler if we don't use it.